aten/src/ATen/native/ForeachUtils.h

```diff
@@ -73,51 +72,63 @@ bool can_use_fast_route(TensorList tensors, Scalar scalar) {
       return false;
     }
 
-    if (t.device() != expected_device) {
-      return false;
-    }
-
     if (!t.is_non_overlapping_and_dense()) {
       return false;
     }
 
-    // complex scalar + integral or boolean tensor will result in complex tensor
-    if (scalar.isComplex() && at::isIntegralType(t.scalar_type(), /*includeBool*/ true)) {
+    if (t.strides() != expected_strides) {
       return false;
     }
+  }
 
-    // float scalar + integral or boolean tensor will result in float tensor
-    if (scalar.isFloatingPoint() && at::isIntegralType(t.scalar_type(), /*includeBool*/ true)) {
-      return false;
-    }
+  return true;
+}
 
-    // integral scalar + boolean tensor will result in integral tensor
-    if (scalar.isIntegral(/*includeBool*/ false) && t.dtype() == at::kBool) {
-      return false;
-    }
+bool tensor_and_scalar_are_same_type(Tensor tensor, Scalar scalar) {
```

Comment:
This isn't really checking if they're the same type, right? A float32 Tensor and a float64 Scalar would pass this check. It's more like it is checking that we don't end up promoting `tensor`. Maybe something like `bool will_promote_tensor(const Tensor& tensor, Scalar scalar)`

Reply:
(also remember to take Tensor by const ref, intrusive_ptr refcount bumps are relatively expensive!)